### ISSUE TITLE
Asynchronous requests with requestAsync() are not actually sent until you call  ->wait()

### ISSUE THREAD

**Issue by drscre** [guzzle/guzzle#1127]
I expected that using `$client->getAsync()` with Curl will fire request immediately, leaving me some time to perform additional work before it completes:

```php
$req = $client->getAsync('http://api.com/get_a_lot_of_data'); // Shouldn't request be sent here ? ...

// do something while the long request is in progress in order not to waste time
do_work();

$req->then(...);
$req->wait(); // ... so here request can already complete
```

But it turns out, that request will actually be fired only when you call `$req->wait()`, acting like a plain synchronous request.

**Comment by Briareos** [comment]
You can check out example no.2 in @mtdowling's response: https://github.com/aws/aws-sdk-php/issues/621#issuecomment-111232418

The key is to invoke `tick` manually instead of relying on `wait` to do so for you. That `Promise::wait` is basically a proxy to `CurlMultiHandler::wait`.

You can call `CurlMultiHandler::tick` once when you enqueue requests via `xxxAsync()`, it won't block. That's because `curl_multi_select` is called first; before any calls to `curl_multi_exec`, that actually fires requests.

To wait for all requests to complete you can use this from the mentioned example:
```php
while (!Promise\is_settled($requests)) {
    $curlMultiHandle->tick();
}
```

**Comment by drscre** [comment]
@Briareos Thanks for explicit `tick()` hint!

Still, I don't quite get whether it's a bug or is it a normal way the things should work?
Why not make `tick()` to get called automatically from `xxxAsync()`?

**Comment by kamilsk** [comment]
> it's a bug or is it a normal way
I think it's a normal way.
> Why not make tick() to get called automatically from xxxAsync()
How you can use `promises` in this case?

**Comment by mtdowling** [comment]
It works this way because calling tick when starting a promise could have the side effect of resolving other promises and resolving them. This could have pretty bad unforeseen consequences-- most of which revolve around unwanted recursion. Take the following example:

You are building a rate-limited pool of concurrent promises that yields promises from a generator. When a promise completes, the abstraction calls next on the generator, sends the request, and adds a promise to the pool. If initiating a promise in cURL caused other promises to complete, then it would cause the abstraction to attempt to call next on the generator while it's still yielding the previous value. This causes a strange sort of recursion in PHP generators and will trigger a fatal error.

You should either use the tick approach I mentioned in the comment linked by @Briareos or use a different adapter that hooks into an event loop in which your whole application is run asynchronously.

**Comment by drscre** [comment]
Unfortunately, calling `tick()` turned out to be not that simple.

In my case I have to execute `tick()` about 2-3 times until the request is actually sent.

As far as I understood, every `while (curl_multi_exec() === CURLM_CALL_MULTI_PERFORM)` loop performs some piece of work on registered handles. 
My conjecture is that you can't say what exactly this piece of work will be. It's up to cURL and, probaly, is subject to changes. 
Maybe first time it will establish connection, next time it will send data, and so on...

So, you can't be sure that your request will be fired on the very first `curl_multi_exec()` loop.

**Comment by Briareos** [comment]
@drscre That's strange; I'm not a cURL expert, but when I look into the ext/curl source code, I see that [`curl_multi_exec`](http://lxr.php.net/xref/PHP_5_6/ext/curl/multi.c#198) is calling [`curl_multi_perform`](http://curl.haxx.se/libcurl/c/curl_multi_perform.html) in libcurl, which says:
>This function handles transfers on all the added handles that need attention in an non-blocking fashion.  
When an application has found out there's data available for the multi_handle or a timeout has elapsed, the application should call this function to read/write whatever there is to read or write right now etc.

My interpretation is that all requests are sent at that point. Are you sure you are calling `tick()` only after all `Client::xxxAsync()` calls?

**Comment by bizmate** [comment]
Hi 
I am also stuck on this, i am sure I missed the point somewhere but it might be beneficial to all if i had my example and maybe troubleshoot for everyone's benefit

```php

use GuzzleHttp\Client;
use GuzzleHttp\Promise;
use GuzzleHttp\Handler\CurlMultiHandler;
use GuzzleHttp\HandlerStack;

...


 public function asyncAction()
    {
        $client = new Client(['base_uri' => '']); //sync client

        $curl = new CurlMultiHandler();
        $handler = HandlerStack::create($curl);
        $client2 = new Client(['handler' => $handler]); // async client

        $start_async = microtime(true);
        $promise = $client2->getAsync('http://host/bigfile.txt');
        $curl->tick();
        $end_async = microtime(true);

        $start_sync = microtime(true);
        $response = $client->request('GET', 'http://host/bigfile.txt');
        $end_sync = microtime(true);

        echo "sync took " . ($end_sync - $start_sync) . ' async ' .  ($end_async - $start_async) ;exit;
    }
```

My objective is to trigger a curl call but dont wait for it even it fails ... just trigger the call and make sure curl has accepted the request task and then exit as fast as possible. I am trying to do this in `$client2` whereas my `$client` is just a standard synchronous call and i am doing a basic benchmark to see how fast an async call is passed to curl.

To verify is the call was executed i just check my server logs for bigfile.txt. At the moment only one call is triggered and recorded in the apache access log, thus confirming the async call is not executed.

What have i missed from @Briareos description?

**Comment by drscre** [comment]
@Briareos

curl_multi_perform() in turn calls [multi_runsingle()](https://github.com/bagder/curl/blob/master/lib/multi.c#L988) on every handle.

multi_runsingle() advances a big state machine.

I look at the source code of [multi_runsingle()](https://github.com/bagder/curl/blob/master/lib/multi.c#L988) and I see some "suspicious" states that can possibly cause first call of multi_runsingle() to return before actually sending http request:
CURLM_STATE_WAITRESOLVE - resolve DNS name
CURLM_STATE_WAITCONNECT - awaiting a completion of an async TCP connect
CURLM_STATE_CONNECT_PEND - no connections, waiting for one

I suspect that a single call to curl_multi_perform() does not guarantee that the http request will be sent during that very call. 
It may just initiate dns resolve or a network connection and return.

**Comment by natebrunette** [comment]
I'm in the same boat as @bizmate.  What I'd like is to know is what <some number of steps> will trigger a non-blocking curl request and allow me to do additional processing.  There could be a number of requests being made, or there could just be one.

For example:

1. Do a network request
2. Do a network request
3. Do a database query
4. Collect the responses of the network requests.

**Comment by cosmolux** [comment]
I'm on the yacht with @natebrunette.  I  see  doing concurrent HTTP requests works easily with guzzleHTTP...
```php
// Wait on all of the requests to complete. Throws a ConnectException
// if any of the requests fail
$results = Promise\unwrap($promises);
```

But I can't get  requests to process asynchronously with other PHP code.  Here is my test case that fails because the 2-second sleep and the 3-second delayed HTTP request add up to 5 seconds, when my hope is they would occur concurrently.  The `sleep 2` represents doing a database query or other task outside of guzzle.

```php
  function testPureGuzzleClientAsyncSleep() {
	$client = new GuzzleHttp\Client();
	
	$startTime = microtime(true);
	$promise = $client->getAsync("httpbin.org/delay/3");
	
	$responseHolder = null;
	$promise->then(function ($response) use (&$responseHolder,$startTime) {
		$this->dump('Got a response! ' . $response->getStatusCode());
		$this->dump("response: ". number_format(microtime(true) - $startTime,4));
		$responseHolder = $response;
	});
	$this->dump("->then() function setup: ". number_format(microtime(true) - $startTime,4));
	$queue = \GuzzleHttp\Promise\queue();
	$queue->run();
	$queue->run();
	$queue->run();
	$queue->run();
	$queue->run();
	$queue->run();
	$queue->run();
	$queue->run();
	$queue->run();
	$this->dump("run run run: ". number_format(microtime(true) - $startTime,4));
	sleep(2);
	$this->dump("sleep: ". number_format(microtime(true) - $startTime,4));
	$promise->wait();
	$this->dump("wait: ". number_format(microtime(true) - $startTime,4));
	
	$results = (string)$responseHolder->getBody();
	$this->dump("getBody: ". number_format(microtime(true) - $startTime,4));
		
	// $this->dump($results);
	$duration = microtime(true) - $startTime;
	
	$this->dump("Total duration: $duration");
	$this->assertTrue($duration < 5);
}
```
Results:
```
  ->then() function setup: 0.0104
  run run run: 0.0110
  sleep: 2.0111
  Got a response! 200
  response: 5.2091
  wait: 5.2092
  getBody: 5.2093
  Total duration: 5.209450006485
```

Am I doing something wrong, or does guzzlehttp not support this fashion of asynchronicity?

**Comment by haikuoyao** [comment]
I'm in the same boat with @cosmolux .
I'm looking for a `fire and forget` request.
I don't care if it fails or not. I just want to send out a request and don't want it to block my code.
I don't need any response.

Can I use guzzle to do it ?

And I try to use RingPhp to do it ? It doesn't work as well. it still blocks my code.

Anyone has any idea ?

**Comment by Dudemullet** [comment]
Has anybody found a good way of doing this? If so would you please post an example? My use case most resembles @haikuoyao 's

>I just want to send out a request and don't want it to block my code. I don't need any response.

**Comment by Dudemullet** [comment]
Ok I got it working cc @haikuoyao. Hopefully this example will help us get everyone of these water vehicles. In all honesty I'm just aggregating and _trying_ to simplify @Briareos and @mtdowling examples into a single one, so thanks to them for the real answers. 

NOTE: This answer is based on using AWS php sdk specifically. You could do it with something else, but it would require to know **what is an HTTP_HANDLER?** and **how do I set it on my library?**.

They key to understanding the fix is in these 2 sentences:

>You can check out example no.2 in @mtdowling's response: aws/aws-sdk-php#621 (comment)

and 

>The key is to invoke tick manually instead of relying on wait to do so for you.

So, basically what they're trying to say here is (me paraphrasing):

>Instead of calling `Promise->wait()` just call `$curl->tick()`, its non-blocking

So... based on this premise if you had code that looked something like:

```php
// just get a promise array from anywhere
$promisesArray = getPromises();

$promisesAggregate = Promise\All($promisesArray);
$result = $promisesAggregate->wait() //forcefully wait
wantToDoThisWithoutWaiting();
```

you could make it non-blocking by changing it to something like:

```php
use GuzzleHttp\Handler\CurlMultiHandler;
use GuzzleHttp\HandlerStack;

$curl = new CurlMultiHandler();
$handler = GuzzleHttp\HandlerStack::create($curl);

$promisesAggregate = Promise\All($promisesArray);
$curl->tick();
yayNoWaiting(); // 🍻
```

## Real code
```php 

$promisesArray = getPromises();

$promisesAggregate = Promise\All($promisesArray);
$curl->tick();

$sdk = new Sdk([
    'http_handler' => $handler,
    'region' => 'us-west-2',
    'version' => 'latest',
]);

$client = $sdk->createS3();
$p1 = $client->listBucketsAsync()->then(function () { echo '-done 1-'; });
$p2 = $client->listBucketsAsync()->then(function () { echo '-done 2-'; });
$aggregate = Promise\all([$p1, $p2]);

foreach($aggregate as $agg){
  $curl->tick(); // put in for loop as one call ddid
}
doSomethingWithoutWaiting();
```

### important stuff
- When initializing your aws sdk, set the `http_handler`!!
- After creating all your promises call `$curl->tick()`

Sorry for this wall of text, hope this can help someone.

**Comment by natebrunette** [comment]
```php
<?php

use GuzzleHttp\Client;
use GuzzleHttp\Handler\CurlMultiHandler;
use GuzzleHttp\HandlerStack;

require_once __DIR__ . '/vendor/autoload.php';

$curl = new CurlMultiHandler(['select_timeout' => 10]);
$handler = HandlerStack::create($curl);
$client = new Client(['handler' => $handler]);

$promise = $client->getAsync('http://httpbin.org/delay/2');

$startTime = microtime(true);
for ($i = 0; $i < 1; $i++) {
    $curl->tick();
}
$duration = microtime(true) - $startTime;
var_dump($duration);

sleep(2);

$startTime = microtime(true);
$promise->wait();
$duration = microtime(true) - $startTime;

var_dump($duration);
```

The above code calls tick once.  This results in:

double(0.00053691864013672)
double(2.2381699085236)

If tick calls is increased to 100, it results in:

double(2.1873440742493)
double(4.6968460083008E-5)

So even if we can trigger the handler with tick, it still seems to block.

@Dudemullet Could you post timing that shows asynchronous requests working correctly with the Amazon SDK?

**Comment by greendrake** [comment]
What a bullshit that asynchronous requests feature! Does not seem to work at all :(.
In this example, the callbacks are never executed unless you uncomment the `wait()` (which of course is not what we want to do):

```
use GuzzleHttp\Client;
use GuzzleHttp\Handler\CurlMultiHandler;
use GuzzleHttp\HandlerStack;
use Psr\Http\Message\ResponseInterface;

$curl = new CurlMultiHandler;
$handler = HandlerStack::create($curl);
$client = new Client(['handler' => $handler]);

$client
    ->getAsync('http://google.com')
    ->then(
        function (ResponseInterface $res) {
            echo 'response: ' . $res->getStatusCode() . PHP_EOL;
        },
        function (\Exception $e) {
            echo $e->getMessage() . PHP_EOL;
        }
    )
    // ->wait()
    ;
$curl->tick();
echo 'bottom' . PHP_EOL;
```
Seems like I'd better use good old `pcntl_fork` and stuff.

**Comment by skerit** [comment]
I can't get this to work, either.
Requests only get sent when calling `wait()`.

Using the `tick()` trick only causes them to block earlier.

**Comment by Briareos** [comment]
I've investigated this further and I'm pretty sure we can't have such asynchronous requests unless someone made a `socket_*` handler that optionally switches to blocking mode during the `sendAsync` request write call. Anything else would require threads or an event loop.

cURL does have to do DNS lookup, so the first `curl_multi_select` call doesn't always send the HTTP request itself. It *might* be possible (I'm not sure) to set the `CURLOPT_RESOLVE` and skip the DNS lookup, but you *do* need to do the lookup yourself then, so... I'd say this issue can't truly be fixed, unfortunately.

**Comment by Briareos** [comment]
@skerit A small debug hint, you can set the `debug` option to `STDOUT` or some file so you can see (in real-time) what each tick does and which one sends the HTTP request itself.

**Comment by undercloud** [comment]
I try to resolve my issue https://github.com/guzzle/guzzle/issues/1650, and it drives me crazy((( ...
`xxxAsync` works or not without `->wait()` and `$curl->tick();` magic? Maybe it works, but not how we expect (like JS\jQuery Promise)?

**Comment by KevinLinggajaya** [comment]
I think I solve the issue, you need to have a running async server setup and worked on your project.
For mine, I setup and configure beanstalkd on my laravel project. I run 
`php artisan queue:listen` 
Guzzle aysncRequest could be execute asynchronously without `$promise->wait()`

**Comment by borkor** [comment]
@KevinLinggajaya 
Could you please elaborate on this?

**Comment by Nokim** [comment]
@greendrake you can't use async without event loop. So, your code should be like this:
```
use GuzzleHttp\Client;
use GuzzleHttp\Handler\CurlMultiHandler;
use GuzzleHttp\HandlerStack;
use Psr\Http\Message\ResponseInterface;

$curl = new CurlMultiHandler;
$handler = HandlerStack::create($curl);
$client = new Client(['handler' => $handler]);

$p = $client
    ->getAsync('http://google.com')
    ->then(
        function (ResponseInterface $res) {
            echo 'response: ' . $res->getStatusCode() . PHP_EOL;
        },
        function (\Exception $e) {
            echo $e->getMessage() . PHP_EOL;
        }
    );
while ($p->getState() === 'pending') {
    $curl->tick();
   //do some other stuff here or just sleep(1);
}
echo 'bottom' . PHP_EOL;
```

**Comment by Nokim** [comment]
The most strange thing is that call to queue->run() doesn't call for tick() inside (with right handler used)
```
            $queue = \GuzzleHttp\Promise\queue();
            $queue->run();
```
But $curl->tick() calls \GuzzleHttp\Promise\queue()->run() inside! And it use delays inside, so yes, @natebrunette, it's blocking but not for long, request itself is much longer.

**Comment by natebrunette** [comment]
@Nokim Yeah, after a lot more research, I realized that because the curl implementation in PHP is blocking, there's no real way to make the requests asynchronously.  What I've resorted to doing is using a pool with callbacks so that one/some of my callbacks has a chance to execute before all of the requests have completed.

**Comment by borkor** [comment]
@natebrunette 
Finally some one said it. It is not async and never will be. There are a whole bunch of closed issues on this subject.

**Comment by alex-shamshurin** [comment]
So what is async here if actual networking  IO call  is blocking? I see `guzzle` is used by some async libs like [http-client](https://github.com/reactphp/http-client/blob/master/src/Request.php) which are for event-loop solution for PHP - ReactPHP and they must be non-blocking by design, but actually they are not -curl is sync and blocking. So what's the use of all this?

**Comment by Nokim** [comment]
@alex-shamshurin You can't say that `guzzle` async call is really non-blocking because it's blocks execution for some small time but you can't say it is really blocking because `guzzle` blocks for much smaller time then request really takes to execute. We are saying about time less then 1 minisecond here.

Main problem of comments above: they missed event loop part. From `guzzle` side there is problem too - it should call `$curl->tick` itself inside of `$query->run()` to hide implementation details (there are other backends except `curl`!). But that has nothing to do with async question.

For example I now run some code which executes 128 requests in parallel. Failed ones take 90 second before timeout but that doesn't block others for noticeable amount of time. Normal requests take usually 2-3 seconds to complete.

But it seems there are some bugs with promises: when I group some async requests with `some` function it blocks execution but it shouldn't. Well, maybe I missed something - promises are new to me.

**Comment by alex-shamshurin** [comment]
Curl sends request and waits for response. It can be 20secs, for example. During this, php thread is blocked by IO. Isn't it?

**Comment by Nokim** [comment]
@alex-shamshurin No, it isn't blocked if you are using promises returned by async request. See my code above. You can do other work waiting request to be done in event loop.

**Comment by alex-shamshurin** [comment]
And what? Promise is promise, it's not an execution part. It's just a state defintions of 'before beginning' and the chain after that. When this promise actually starts its execution (via tick or wait) it will start actual request which is blocking by design. It it wiil block the whole loop, because it's sync.  No matter how you organize you code - callbacks, promises, generators - it's just a style. Actual performing must be async as well.  Bit curl does not support that.

**Comment by Nokim** [comment]
@alex-shamshurin You are really wrong here. Otherwise whole concept of JavaScript (i.e. Node.js) will fail to work. Promises are means to easily break code in small pieces and execute them when data available. Event loop is essential here, in it you check if data available and call right piece of code. Data can be delivered by OS kernel itself into buffers so no real need for threading or additional process inside program, just check buffers on regular basis. That's what curl does - handles all that complicated things under hood and can do so in async way too. `guzzle` and promises simplifies things even further.

Well, we are really offtopic here...

**Comment by alex-shamshurin** [comment]
NodeJS has a lot of async functions, actually all of them are async, except those which contain `sync` suffix in the name, like `fs.readFileSync`. At C++ level they run in separate threads and when they finished their tasks they emit events that task is finished. This means that JS code with Event Loop itself is inside one thread, but actual NodeJS `readFile` and other IO functions run in another thread. So NodeJS is actually multithreaded, one thread is waiting via `epoll_wait` for file events, another is performing Event Loop and callbacks processing and is free when you read big file by another thread. (you can easily check it with pstack or GDB) 

Instead, PHP Curl extension is not dispatched to a separate thread because PHP itself is not multithreaded, it's just a call to C functions and PHP process have to wait until it finished. The same is for `file_get_contents` and other functions. See [here](https://github.com/reactphp/react/wiki/FAQ)

**Comment by jollaman** [comment]
I've been struggling a couple days trying to understand how to make these calls async. And finally I believe I did it. So for anybody looking for a simple example:

 1 - Create a method which returns a promise for each service/API you want to call, like this
`function getAPI1Promise($handler){
	
	$client = new Client([
    'timeout'  => 15.0,
	'handler' => $handler
	]);
	
	$promise = $client->requestAsync('GET', "https://www.myapiurl1.com/whaveruwanttocall");

	return $promise;
}

function getAPI2Promise($handler){
	
	$client = new Client([
    'timeout'  => 15.0,
	'handler' => $handler
	]);
	
	$promise = $client->requestAsync('POST', "https://www.myapiurl2.com/whaveruwanttocall", ['json' => ['parameter' => 'stuff']]);

	return $promise;
}`

2 - Within your main code create a multihandler, enqueue promises and call "tick"
`
$multicurl = new CurlMultiHandler;
$multihandler = HandlerStack::create($multicurl);
	
$promise1 = getAPI1Promise($multihandler);
$promise2 = getAPI2Promise($multihandler);

$multicurl->tick();
`

3 - Append any other code you want to run while all requests are being executed
4 - Handle each response when you need it

```
$response1 = $promise1->wait();
$response2 = $promise2->wait();
		
$body = (string)($response1->getBody());
...
```

All requests will be executed concurrently. 

Bye

**Comment by natebrunette** [comment]
@jollaman How are you verifying this behavior? From what I understand, curl in php is a synchronous operation. I don't think `$multicurl->tick();` will start the requests, but if it does, it should block execution until it gets a response.

I've found making requests to httpbin through something like http://httpbin.org/delay/3 is a good way to test this.

**Comment by jollaman** [comment]
I'd say "I don't know why, but it works". Anyway it is using curl_multi... functions which allow for concurrent requests. It seems to me it's sending all requests altogether and waiting exactly as I want it to do. My page took around 1 minute to complete as I was calling around 15 different API's and some are answering in 3/4 seconds or just time-out after 15 s. Now it takes just 16-17 s and that's because 1 of them is down at present. So it works...

**Comment by natebrunette** [comment]
I don't think the confusion has ever been about how to make multiple concurrent requests. That should work without too much hassle by calling `sendAsync()` and then calling wait on each promise.

From what I've seen, the confusion is around how to do that _and_ do additional processing in PHP while the requests are being made, which isn't possible using only PHP.

**Comment by jollaman** [comment]
sendAsync+wait results in synchronous, blocking code without "tick". And my page is actually doing other tasks in the middle (I added timestamps to check how it behaves), I'm just waiting for answers almost at the end of the code where I absolutely need them. So everything seems to be concurrent to me...

**Comment by jollaman** [comment]
Don't know what to say, just seems to work to me. BTW, I measured the time before and after the call to "tick", and it lasts no more than 200ms/300ms so no, it's not blocking.

**Comment by natebrunette** [comment]
If you add a sleep for 15-30 seconds and then call wait, does it still wait an additional 15 seconds before resolving the requests? My guess is yes. I'm not saying there's no benefit to making concurrent requests, but this issue is about making asynchronous requests.

**Comment by jollaman** [comment]
Nope, it's waiting around 5 s more (22s instead of 17s), and that's because I still have sync calls in the main code, which I didn't rework yet. Well the original thread ( @drscre ) was about concurrent requests...

**Comment by shtse8** [comment]
I have this problem too. I keep sending async request using Guzzle, but I found it is blocked a few milliseconds. 
```
Loop::repeat(0, \Closure::bind(function () use (&$timer) {
				if (!empty($this->handles)) {
					$this->tick();
				}
			}, $handler, $handler));
```

But I tried to use https://github.com/amphp/artax to do async request.
They haven't any problem because they are implementing the request themselves, but resulting using too much CPU.
How can I make it work properly using Guzzle Curl?

My Code; ( I'm using Amphp, not React, but it should be the same.)
```
$guzzleClient = null;
function request($url, $method, $options) : Promise
{
	global $guzzleClient;
	
	$deferred = new Deferred;
	
	/* Create Client - Singleton */
	if (!$guzzleClient) {
		$defaultRequestHeaders = [];
		$defaultRequestHeaders['User-Agent'] = 'Mozilla/5.0 (Windows NT 6.1; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/39.0.2171.71 Safari/537.36';
		
		$handler = new CurlMultiHandler();
		Loop::repeat(0, function() use ($handler) {
			$handler->tick();
		});
		$guzzleClient = new GuzzleClient([
			'headers' => $defaultRequestHeaders,
			'handler' => HandlerStack::create($handler),
			'connect_timeout' => 5,
			'read_timeout' => 5,
			'timeout' => 5
		]);
	}
	
	/* Send Request */
	$guzzleClient
		->requestAsync($method, $url, $options)
		->then(function($response) use ($deferred) {
			$deferred->resolve($response);
		})
		->otherwise(function($e) use ($deferred) {
			if ($e instanceof Throwable) {
				$deferred->fail($e);
			} else {
				$deferred->fail(new Exception($e));
			}
		});
		
	return $deferred->promise();
}
```

**Comment by mtdowling** [comment]
Asynchronous does not mean non-blocking. The curl handler in Guzzle allows asynchronous requests but does block.

And no, you can't really mix event loops. Using The curl handler with Artax or React is not a good idea because the curl calls will block. You can, however, create or use a custom handler for interop with these projects so that you can use Guzzle in a non-blocking way.

**Comment by shtse8** [comment]
Got it, thanks. so what handler should I use to use Guzzle in a non-blocking way? (I have a command line GUI, I need to print the screen > 10 times every second, so blocking is not acceptable in my project)

**Comment by alex-shamshurin** [comment]
Guys, there is really no way in PHP to make IO calls non-blocking. You can organize your code in async way, but each IO functrion will block PHP itself. So it useless to use PHP in this case.

**Comment by shtse8** [comment]
But how can amphp/artax do so well? I am using it without any blocking happen. but it eats lots of cpu when many requests.

**Comment by alex-shamshurin** [comment]
NodeJS uses a separate thread to perform its IO opearations and ONE thread to run event loop and uses native threaded cocurrency. Artax use only one PHP thread (there is always one) for webserver itself and for polling network sockets from time to time and it implements ONE THREAD concurrency. It does not block for seconds (minutes) like curl does, instead it works at socket level and switching from socket to socket. Anyways it's a single threaded solution! 

Say, if you download two files simultaneously, it will switch from one to another socket and to EV itself and do it in ONE thread. It just devides its job into small parts of data. So it's async but not parallel.

**Comment by shtse8** [comment]
Yes, I understand the threading problem. For the case just sending requests, not downloading files, using single thread is far more enough. But I still don't understand why Curl needs to block for seconds for async, while Artax doesn't need.

**Comment by alex-shamshurin** [comment]
Because `Curl` reads data from socket until it finished and only then returns control to your PHP code.  AMP blocks PHP at `Loop::run` string and then all magic is done inside - it reads data from sockets and switches between them and EL. But it STOPS your code at `Loop:run` and also does not return control until it stops somehow or by exception.

Meanwhile, APM can implements TRUE multithreading concurrency if PHP itself compiled with [phtreads](https://pecl.php.net/package/pthreads) and you use [parallel](https://github.com/amphp/parallel) for AMP

**Comment by halaei** [comment]
I don't know about node.js architecture, but I am sure multi-threading is not required for non-blocking I/O. Maybe some threading is involved in OS kernel, or whatever, but everything in high-level is doable via a single thread. Here is how it can be done with php:
http://php.net/manual/en/function.stream-set-blocking.php
http://php.net/manual/en/function.socket-select.php
http://php.net/manual/en/function.stream-select.php

What is easy to do with node.js but hard with php, is maybe to wait for all the events from timers to files and socket and http connections in a single loop wrapping a call to the `select ` system function?
Other than that, I will be happy if someone tell me what PHP really lacks?

**Comment by alex-shamshurin** [comment]
@halaei Do you really understand how concurrency may be achieved? For anything that CPU does it must be some thread - this one or another one. You cannot send you job "somewhere" and magically get a result later. It must be a CPU job run at some thread or at some other process (fork). So only one way to perform some tasks in parallel (simultaneously) is to dispatch a job to a separate thread. If you write a function or library which read some bytes from a file and then switches to read some bytes from another file - it's not a real concurrency and this is what AMP/Artax does.  It does only one thing simultaneously - one by one, just swtiching between them. On the contrary, NodeJS's all 'readFile' or others long-run operations run in their own thread. So non blocking IO means keeping MAIN LOOP non blocked in one thread, but if IO start do something long in a main thread - it will be blocked. 

So APM, ReactPHP  are blocking by nature if it does not use phtreads. But anyway it can gain performance because socket recieves data from time to time (packets are not available all the time) and EV can switch to another socket, where data is availbale already and read from there.

"kernel level" is a thread iself, it usually has not been created by fork() or clone() system calls. All IO operations are in "user level".

**Comment by alex-shamshurin** [comment]
> PHP really lacks
language itself, single threaded (phtreads? no..), OOP, no JIT, awful GC.
 Do not forget - PHP is interpreter, but JS code is compiled into machine code.

**Comment by halaei** [comment]
> Do you really understand how concurrency may be achieved?
Yes. And I have done multi-threading in C/C# for Linux/windows. But multi-threading and forking is not all the mechanisms for concurrency when we are talking about I/O. For example, not all the job of sending and receiving data from disk/network into main memory is handled by CPU. The corresponding hardware just sends data into memory without CPU being involved, and when they are done, they just generate an interrupt, so some OS code will run and process the raw input and does the rest, dealing with sockets and file handlers and .... That is called "direct memory access", a concept that I was in direct contact with. When the data is inside the memory and ready to be processed, the remaining job can be done so quickly, relative to the hardware delay, that you don't even need a separate thread for it.

And if there should be some threading for another reason I am not aware of, I think that is done inside the `select` function by some linux developer or inside `socket_select` by some core php developer. If they forget to do so, somebody just report it or send them a pull request.

> "PHP reaally lacks" single threaded (phtreads? no..), OOP, no JIT, awful GC.

1. PHP has a pthreads extension, if you so miss multi-threading just use it.
2. You can write object oriented code in PHP as you can do in C++. Someone might write just shit in a purely object oriented language like Java, and someone can write beautiful, even object oriented code in a language that is not object oriented, like C. And PHP is way more object oriented than JavaScript.
3.  No JIT. I am fine with it as long as I don't forget to enable opcache, because I don't do my heavy CPU-bound tasks with PHP. You may try HHVM or wait for PHP 8.0, or just not use PHP, but that has nothing to do with our discussion!
4. awful GC? Do you have any reference/benchmark for that? It have worked perfectly fine for me.

**Comment by alex-shamshurin** [comment]
> "direct memory access"

Agree with all

> PHP has a pthreads extension, if you so miss multi-threading just use it.

Agree, but it's not for windows platform. Many developers still use windows.. I really do not think it's like Java or C# level solution. I do not know people who really use it. 

> PHP is way more object oriented than JavaScript

In JS world we have all from OOP but we can use much better things. JS much more powerful. 

> awful GC?

PHP always has memory leaks. Still. It can't runnig for months. 

It's a little offtop I think.

**Comment by shtse8** [comment]
> PHP always has memory leaks. Still. It can't runnig for months.

It's true.

**Comment by halaei** [comment]
Memory leak is tolerable for no programming language, so I highly doubt that there can be any memory leak in PHP interpreter that no one has ever cared to detect and fix. I don't know, maybe php 5.4 but not php 7.1.

And if there is a memory-leak bug in PHP 7.1, it is not the fault of PHP itself, it is the fault of us, PHP developers, who has not yet reported it.

I can imagine the chance of an undiscovered leak in some PHP library/framework to be higher than a leak in some JS library with equivalent popularity. My reason is that JS libraries are used more often in the long-running applications, so the memory related bugs will be discovered faster.

**Comment by alex-shamshurin** [comment]
I cannot explain why, but it's a fact. Our backend developer made a REST (with quering our data provider via curl) on ReactPHP and it leaks on 7.1. I wrote some PHP long-running workers for image files processing some years ago and have encountered with same memory leaks. I think PHP was born to die (after request finished) and it's his nature by design. Now I use Java, Python or NodeJS for backend with multithreading and never got these problems.

**Comment by shtse8** [comment]
So how can I do the async requesting without blocking using Guzzle?

**Comment by alex-shamshurin** [comment]
No. It uses curl and it always blocks.

**Comment by kelunik** [comment]
@alex-shamshurin I've been running some processes for months without any issues, it's definitely possible.

**Comment by alex-shamshurin** [comment]
@kelunik Please read the full thread above, It's totally diffrent in how it uses IO. NodeJS is single-threaded only in EV and its callbacks. If you read file async or make a DB query it runs on a separate thread. Use a debugger to check it out. PHP will block its EV thread until this IO operation finished. Do not let them fool you.

**Comment by halaei** [comment]
This discussion is killing me! I read the libuv documentation (https://media.readthedocs.org/pdf/libuv/stable/libuv.pdf). There are 2 approaches to do async I/O:
1. Doing blocking I/O inside a sepearate thread, like @alex-shamshurin mentioned.
2. Using platform specific non-blocking APIs inside a single thread, as I mentioned.

And here is what libuv does:
> libuv uses a thread pool to make asynchronous file I/O operations possible, but network I/O is
always performed in a single thread, each loop’s thread.

It means libuv decides to use the first approach for file I/O and the second for network I/O. The reason for them to chose the first approach for file I/O is explained here (http://blog.libtorrent.org/2012/10/asynchronous-disk-io/).

**Comment by shtse8** [comment]
I don't familiar with lib extension. Does it mean we can make use of libuv to make PHP more powerful for asynchronous I/O and better event loop?

**Comment by halaei** [comment]
@shtse8 Similar libraries are ported to PHP as extensions. [reactphp](https://github.com/reactphp/react/blob/master/composer.json) can use one of the followings:
```json
"suggest": {
        "ext-libevent": "Allows for use of a more performant event-loop implementation.",
        "ext-libev": "Allows for use of a more performant event-loop implementation.",
        "ext-event": "Allows for use of a more performant event-loop implementation."
}
```
Take a look at http://reactphp.org/event-loop/#loop-implementations

**Comment by kelunik** [comment]
@shtse8 Yes, exactly. Amp can use `libuv` just like Node. It will use `libuv`'s file I/O threading if available. See https://amphp.org/amp/event-loop/#implementations.

Extensions allow other loop backends that can handle more file descriptors as `stream_select`, which is limited to 1024 FDs usually.

**Comment by shtse8** [comment]
Does it mean using libuv instead of libev can get better performance without changing anything in code?

**Comment by shtse8** [comment]
https://github.com/amphp/amp/blob/5b2f54707ca5d6d1e541ceeafa8b4904e5ea4837/lib/Loop/DriverFactory.php#L19
It seems libuv is better than libev, as amphp prefer to use libuv first.

**Comment by kelunik** [comment]
@shtse8 It won't bring (much) performance benefits, but allows using `libuv` for file IO instead of multiple processes.

**Comment by alex-shamshurin** [comment]
Guy, we can try all them, but I'm sure you cannot do blocking by nature calls in single-threaded environment.  Read what ReactPHP FAQ https://github.com/reactphp/react/wiki/FAQ says:

> How can I use blocking functions?

The short version is: **You can't.**

ReactPHP is based on non-blocking I/O. Since PHP is single-threaded, that means that you cannot make any blocking calls without blocking the entire process. Which means that if you have 1K concurrent users, if each one blocks for just one millisecond, it will block the entire process for a whole second.

What are blocking functions? Pretty much any existing PHP code that touches the filesystem or the network. This includes, but is not limited to:

sleep
file_get_contents
file_put_contents
mkdir
fsockopen
PDO
mysql_*
etc.

**Comment by bimbojones** [comment]
If someone needs it.

`
$client = new Client();
$promises = [];
foreach($data as $value)
    $promises[] = $client->requestAsync('GET', $url . '/' . $value)
                ->then(function($response) {
                       // do something
                });

$results = Promise\unwrap($promises);`

This makes multiple calls

**Comment by quangmach** [comment]
@bimbojones This one works! Thank you!

**Comment by songokjesse** [comment]
> $client = new Client();
> $promises = [];
> foreach($data as $value)
> $promises[] = $client->requestAsync('GET', $url . '/' . $value)
> ->then(function($response) {
> return json_decode($response->getBody(), true); //Just added this for getting the response
> });
> 
> $results = Promise\unwrap($promises);

I have been struggling with guzzle promise in laravel and this actually worked for me

**Comment by pixagraphic** [comment]
I had the same issue. I wanted to fire asynchronous post requests to a remote server in my laravel app using guzzle without blocking any other code on multiple occasions. What I finally did is writing a wraper for the guzzle request into a laravel "job". Now I can fire requests using MyJob::dispatch(dataArray); when laravel is configured correctly this gives true asynchronous requests that do not block.

Make sure to set QUEUE_CONNECTION=database in your .env file and to run "artisan queue:table" and "artisan migrate". Eventually you need to make sure that the queue worker is running. "artisan queue:work" The worker must be kept running by using a cronjob for example. (Refer to: https://gist.github.com/codeInBit/e900538d338747959ac48515fc22cc76)

**Comment by aarongtz** [comment]
> @greendrake you can't use async without event loop. So, your code should be like this:
> 
> ```
> use GuzzleHttp\Client;
> use GuzzleHttp\Handler\CurlMultiHandler;
> use GuzzleHttp\HandlerStack;
> use Psr\Http\Message\ResponseInterface;
> 
> $curl = new CurlMultiHandler;
> $handler = HandlerStack::create($curl);
> $client = new Client(['handler' => $handler]);
> 
> $p = $client
>     ->getAsync('http://google.com')
>     ->then(
>         function (ResponseInterface $res) {
>             echo 'response: ' . $res->getStatusCode() . PHP_EOL;
>         },
>         function (\Exception $e) {
>             echo $e->getMessage() . PHP_EOL;
>         }
>     );
> while ($p->getState() === 'pending') {
>     $curl->tick();
>    //do some other stuff here or just sleep(1);
> }
> echo 'bottom' . PHP_EOL;
> ```

I don't get why you need to iterate over the promise state, could you explain, please?, this isn't mentioned in guzzle's documentation's async part as far as I'm aware.

**Comment by Nokim** [comment]
> @Nokim I don't get why you need to iterate over the promise state, could you explain, please?, this isn't mentioned in guzzle's documentation's async part as far as I'm aware.

Basically that's what `$p->wait()` does. You have to wait until promise resolves and you have to do implicit call to `$curl->tick()` meanwhile if you are using curl with Guzzle to let things to move on. I'm not using `$p->wait()` here because it doesn't that call (it's backend dependent). If you have long-running process with many regular calls to Guzzle it will work on its own because of built-in calls, but in one-shot call you have to do it yourself.

I've come to this reading Guzzle's source code. I don't work with it anymore so I don't know how it is is in newer versions. I don't remember why async call was used here instead of sync one, but I'm sure there was a reason.

Oh, and you have to iterate over promise state just because otherwise process will move on and terminate before you get result. That's the whole point of async code - you can do other things meanwhile. End of program included.

**Comment by KNobHack** [comment]
I found [this](http://www.onlineaspect.com/2009/01/26/how-to-use-curl_multi-without-blocking/) old blog post about non blocking curl_multi. I have tested his cod in php 7.4 and suprisingly working. Could we implement this?
